### ISSUE TITLE
Segmentation fault in CV macro atom spectral cycle

### ISSUE THREAD

**Issue by jhmatthews** [sirocco-rt/sirocco#59]
**Status:** Unresolved
**Version:** 76c+
**Description:** Segmentation fault after the macro atom level emissivities have been calculated in CV models.
**Model:** SV93 CV model, macro atom line transfer (mode 7), mass loss rate >=1e-9 msol/yr, ionization mode 3.

I've only observed this problem in parallel mode, which makes me think it is something to do with my parallelisation of macro atom level emissivities. It also see

Tricky do debug, due to CV performance problems (#54), but I'm working on it. It seems to also be isolated to high density cases, which is exceedingly odd...something which gets overflowed only if the density is high i.e. number of scatters or something similar? 

I will do another test run and post error messages here. Should not affect anyone else's work and has only been observed with this specific case.

**Comment by jhmatthews** [comment]
A little more info on this: I can confirm that this definitely happens once the macro atom emissivities have been calculated. Here's the error message:

```
Totals: f_matom 4.713042e+33 f_kpkt 1.116280e+33
photon_gen: band 1.62e+15 to 3.53e+15 weight 1.75e+29 nphotons 25000 ndisk  17869 nwind      0 nstar    477 npow 0 
photo_gen_kpkt creates nphot  1274 photons from 18346 to 19620 
photo_gen_matom creates nphot  5380 photons from 19620 to 25000 
[jamess-imac:35797] *** Process received signal ***
[jamess-imac:35797] Signal: Segmentation fault: 11 (11)
[jamess-imac:35797] Signal code: Address not mapped (1)
[jamess-imac:35797] Failing at address: 0x7ff1ecfb9d40
[jamess-imac:35797] [ 0] 2   libsystem_c.dylib                   0x00007fff8a1f48ea _sigtramp + 26
[jamess-imac:35797] [ 1] 3   ???                                 0x3ba05f4a3c5e5e39 0x0 + 4296538816956489273
[jamess-imac:35797] [ 2] 4   py76c_dev                           0x000000010d4968d0 xmake_phot + 1152
[jamess-imac:35797] [ 3] 5   py76c_dev                           0x000000010d495eec define_phot + 188
[jamess-imac:35797] [ 4] 6   py76c_dev                           0x000000010d48f389 main + 13577
[jamess-imac:35797] [ 5] 7   libdyld.dylib                       0x00007fff910fa7e1 start + 0
[jamess-imac:35797] *** End of error message ***
--------------------------------------------------------------------------
mpirun noticed that process rank 0 with PID 35797 on node jamess-imac.phys.soton.ac.uk exited on signal 11 (Segmentation fault: 11).
--------------------------------------------------------------------------
jamess-imac:template jmatthews$ 
```

I'm testing if this occurs in h20 only models (no He) and in non-parallel mode...although the non-parallel test may take a while!

**Comment by kslong** [comment]
It would be interesting to know if this means it is failing in line 1152 of xmake_phot.  I presume so.  Knox

**Comment by jhmatthews** [comment]
that's what I'd thought, there isn't a line 1152 in xmake_phot though. Pretty sure it's going wrong somewhere in photo_gen_matom.

131128: Have now isolated this to line 2696 in photo_gen_matom (matom.c) [for future reference, I think the 1152 refers to the number of bytes or characters in machine code or something similar, not in the source file itself]:

``` C
  photstop = photstart + nphot;
  Log ("photo_gen_matom creates nphot %5d photons from %5d to %5d \n", nphot,
       photstart, photstop);

  for (n = photstart; n < photstop; n++)
    {
      /* locate the wind_cell in which the photon bundle originates. And also decide which of the macro
         atom levels will be sampled (identify that level as "upper"). */

      xlum = (rand () + 0.5) / (MAXRAND) * geo.f_matom;

      xlumsum = 0;
      icell = 0;
      upper = 0;
      while (xlumsum < xlum)
    {
      if (wmain[icell].vol > 0.0)
        {
          nplasma = wmain[icell].nplasma;         <<<<< SEG FAULT<<<
          xlumsum += macromain[nplasma].matom_emiss[upper];
          upper++;

```

Working out why this occurs now, but it would appear to be something to do with the sum over the PDF running off the end of the wind structure, probably due to an error in my parallelisation of the emissivities.

**Comment by jhmatthews** [comment]
Turns out when I parallelised the emissivity calculation I made a mistake. Once all the information has been calculated and broadcast, then there is a little loop which works out geo.f_matom by going over all cells, so it is not just the contribution from that thread:

``` C
  /* this next loop just corrects lum to be the correct summed value in parallel mode */
  lum = 0.0;   // need to rezero, fixes segfault bug #59 

  for (n = 0; n < NPLASMA; n++)
    {

      for (mm = 0; mm < nlevels_macro; mm++)
    {
      lum += macromain[n].matom_emiss[mm];
    }
    }
```

I'd forgotton to have the line lum = 0.0, which meant that the geo.f_matom was overly high and when it went through the photo_gen_matom routine it couldn't find a cell for the photon to originate from and ran off the end of the array. All fixed now, and runs now finish, Hurray!
